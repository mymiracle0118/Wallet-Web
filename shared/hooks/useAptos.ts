import create from 'zustand'
import { devtools, persist } from 'zustand/middleware'
import { Ed25519Keypair } from '@mysten/sui.js/keypairs/ed25519'
import { chromeSyncStorage } from '@portal/shared/utils/chromeSyncStorage'
import { getAptosWalletUsingSeed, getBalance } from '@portal/shared/services/aptosService'
import { ethers } from 'ethers'
import { toUTF8Array } from '../utils/utf8'

type APTOSWalletState = {
  /** APTOS network - create wallet */
  createAptosWallet: (mnemonics?: string) => Promise<any>
  /** APTOS network - Update balance */
  refreshBalance: (aptosType?: string) => Promise<void>
  _keypair?: Ed25519Keypair | null
  /** APTOS network - Wallet Address */
  address?: string
  /** APTOS network - privateKey Address */
  privateKey?: string
  /** APTOS network - Total Balance of wallet */
  balance: number
  /** APTOS network - All objects owned by wallet */
  objectsOwned?: Array<any>

  wallet?: ethers.Wallet
  encryptedPrivateKey: (string | number)[]
  encryptedWallet?: string
}

const initialState = {
  address: undefined,
  balance: 0,
  _keypair: undefined,
  privateKey: undefined,
  encryptedPrivateKey: [],
}

export const useAptos = create<APTOSWalletState>()(
  devtools(
    persist(
      (set, get) => ({
        ...initialState,

        createAptosWallet: async (mnemonics?: string) => {
          const wallet = await getAptosWalletUsingSeed(mnemonics)
          if (wallet) {
            const encryptedPrivateKey = toUTF8Array(wallet?.privateKey)
            set({ address: wallet?.address, encryptedPrivateKey })
          }

          const { refreshBalance } = get()
          await refreshBalance()
          return wallet
        },
        refreshBalance: async (aptosType?: string) => {
          const { address } = get()
          if (!address) return
          const balance = await getBalance(address, aptosType)
          set({ balance: balance })
        },
      }),
      {
        name: 'APTOS-storage',
        getStorage: () => chromeSyncStorage,
        partialize: (state) => Object.fromEntries(Object.entries(state).filter(([key]) => !['wallet'].includes(key))), // Prevents decrypted wallet from being persisted in storage
      }
    ),
    { name: 'APTOS - Wallet' }
  )
)
