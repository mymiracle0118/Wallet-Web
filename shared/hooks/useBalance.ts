import { useEffect, useMemo } from 'react'
import { useQuery } from '@tanstack/react-query'
import { BigNumber } from 'ethers'

import { NetworkFactory } from '@portal/shared/factory/network.factory'
import { NetworkToken } from '../utils/types'
import { useStore } from './useStore'

type BalanceResponse = {
  balance: BigNumber | number
  balanceFormatted: number
  isLoading: boolean
}
const { updateTokenBalance } = useStore.getState()
export const useBalance = (token: NetworkToken): BalanceResponse => {
  let balance, balanceFormatted

  const networkFactory = NetworkFactory.selectByNetworkId(token.shortName)
  if (!networkFactory) {
    return {
      balance: 0,
      balanceFormatted: 0,
      isLoading: false,
    }
  }

  const walletAddress = token.address
  const { data, isLoading } = useQuery(
    ['asset-balance', token.networkName, walletAddress, token.shortName],
    () => networkFactory.getTokenBalance(walletAddress as string, token),
    {
      refetchInterval: 60000,
      staleTime: 30000,
    }
  )

  balance = useMemo(() => {
    return data || BigNumber.from(0)
  }, [data])
  balanceFormatted = useMemo(() => networkFactory.getBalanceFormattedValue(data), [data]) as number

  useEffect(() => {
    // Update asset balance in wallet to keep running total
    if (balance && balanceFormatted) {
      updateTokenBalance(token.shortName, balance, balanceFormatted)
    }
  }, [balance, balanceFormatted])

  return {
    balance,
    balanceFormatted,
    isLoading,
  }
}
