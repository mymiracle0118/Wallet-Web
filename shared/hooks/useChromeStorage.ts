import { useCallback, useEffect, useState } from 'react'

export const storage = {
  get: (key: string, defaultValue: any, storageArea: Storage) => {
    const keyObj = defaultValue === undefined ? key : { [key]: defaultValue }
    return new Promise((resolve, reject) => {
      chrome.storage[storageArea].get(keyObj, (items: any) => {
        const error = chrome.runtime.lastError
        if (error) return reject(error)
        resolve(items[key])
      })
    })
  },
  set: (key: string, value: any, storageArea: Storage) => {
    return new Promise((resolve, reject) => {
      chrome.storage[storageArea].set({ [key]: value }, () => {
        const error = chrome.runtime.lastError
        error ? reject(error) : resolve()
      })
    })
  },
}

/**
 * Basic hook for storage
 * @param {string} key
 * @param {*} initialValue
 * @param {'local'|'sync'} storageArea
 * @returns {[*, function(*= any): void, boolean, string]}
 */

export default function useChromeStorage(key, initialValue, storageArea) {
  const [INITIAL_VALUE] = useState(() => {
    return typeof initialValue === 'function' ? initialValue() : initialValue
  })
  const [STORAGE_AREA] = useState(storageArea)
  const [state, setState] = useState(INITIAL_VALUE)
  const [isPersistent, setIsPersistent] = useState(true)
  const [error, setError] = useState('')

  useEffect(() => {
    storage
      .get(key, INITIAL_VALUE, STORAGE_AREA)
      .then((res) => {
        setState(res)
        setIsPersistent(true)
        setError('')
      })
      .catch((error) => {
        setIsPersistent(false)
        setError(error)
      })
  }, [key, INITIAL_VALUE, STORAGE_AREA])

  const updateValue = useCallback(
    (newValue) => {
      const toStore = typeof newValue === 'function' ? newValue(state) : newValue
      storage
        .set(key, toStore, STORAGE_AREA)
        .then(() => {
          setIsPersistent(true)
          setError('')
        })
        .catch((error) => {
          // set newValue to local state because chrome.storage.onChanged won't be fired in error case
          setState(toStore)
          setIsPersistent(false)
          setError(error)
        })
    },
    [STORAGE_AREA, key, state]
  )

  useEffect(() => {
    const onChange = (changes, areaName) => {
      if (areaName === STORAGE_AREA && key in changes) {
        setState(changes[key].newValue)
        setIsPersistent(true)
        setError('')
      }
    }
    chrome.storage.onChanged.addListener(onChange)
    return () => {
      chrome.storage.onChanged.removeListener(onChange)
    }
  }, [key, STORAGE_AREA])

  return [state, updateValue, isPersistent, error]
}
