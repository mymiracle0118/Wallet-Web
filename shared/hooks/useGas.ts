import React, { useEffect, useState } from 'react'
import { useQuery } from '@tanstack/react-query'
import { ethers } from 'ethers'
import { getContractInterface } from '@portal/shared/utils/getContractInterface'
import { NetworkFactory } from '@portal/shared/factory/network.factory'
import { Asset, GasOption, NetworkToken } from '@portal/shared/utils/types'

const getERC20GasLimit = async (asset: Asset, provider: any): Promise<number> => {
  if (asset.type === 'erc721') {
    return 100000
  }
  const contract = new ethers.Contract(asset.contractAddress!, getContractInterface(asset.type), provider)
  const deploymentData = contract.interface.encodeDeploy([])
  const estimatedGasLimit = await provider.estimateGas({ data: deploymentData })
  return Number(ethers.utils.formatUnits(estimatedGasLimit, 'wei'))
}

export const useGas = (asset: NetworkToken, option?: GasOption) => {
  const [gasLimit, setGasLimit] = useState(21000)
  const [gasOption, setGasOption] = useState(option || GasOption.Aggressive)
  const getNetwork = NetworkFactory.selectByNetworkId(asset.networkName)
  const provider = getNetwork.provider
  const { data: feeData } = useQuery(
    ['gas-fee-data', asset],
    () => {
      return getNetwork.getFeeData()
    },
    {
      staleTime: 15000,
      refetchInterval: 15000,
    }
  )

  const { data: gasLimitData } = useQuery(
    ['gas-limit-data', asset, provider],
    () => getERC20GasLimit(asset, provider),
    {
      enabled: asset.tokenContractAddress !== undefined,
      staleTime: 60 * 60000,
    }
  )

  useEffect(() => {
    if (gasLimitData) {
      setGasLimit(gasLimitData)
    }
  }, [gasLimitData])

  const maxFeePerGas = React.useMemo(() => {
    if (!feeData || feeData.gasPrice === null) {
      return ethers.utils.parseUnits('0', 'gwei')
    }

    let baseGas = parseFloat(ethers.utils.formatUnits(feeData.gasPrice || 0, 'gwei'))
    const baseGasPlusMultiplier = (multiplier: number) => {
      return baseGas * multiplier
    }

    switch (gasOption) {
      case GasOption.Low:
        baseGas = baseGasPlusMultiplier(1.0)
        break
      case GasOption.Market:
        baseGas = baseGasPlusMultiplier(1.5)
        break
      case GasOption.Aggressive:
        baseGas = baseGasPlusMultiplier(2.0)
        break
    }

    return ethers.utils.parseUnits(baseGas.toFixed(8), 'gwei')
  }, [gasOption, feeData])

  const maxPriorityFeePerGas = React.useMemo(() => {
    return getNetwork.getMaxPriorityFeePerGas({ maxFeePerGas, gasOption })
  }, [gasOption, maxFeePerGas])

  const marketGas = React.useMemo(async () => {
    return await getNetwork.marketGas(maxFeePerGas)
  }, [gasOption, maxFeePerGas])

  const estimatedTransactionCost = React.useMemo(async () => {
    return maxFeePerGas && (await getNetwork.getEstimatedTransactionCost(maxFeePerGas, maxPriorityFeePerGas, gasLimit))
  }, [gasLimit, gasOption, maxFeePerGas])

  const estimatedTransactionTime = React.useMemo(() => {
    switch (gasOption) {
      case GasOption.Low:
        return '30 seconds'
      case GasOption.Market:
        return '30 seconds'
      case GasOption.Aggressive:
        return '15 seconds'
    }
  }, [gasOption])

  return {
    gasLimit,
    gasOption,
    setGasOption,
    marketGas,
    maxFeePerGas,
    maxPriorityFeePerGas,
    estimatedTransactionCost,
    estimatedTransactionTime,
  }
}
