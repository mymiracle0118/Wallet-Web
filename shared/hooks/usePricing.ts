import { useQuery } from '@tanstack/react-query'
import { utils } from 'ethers'
import React from 'react'

import { getSimplePrice } from '@portal/shared/services/coingecko'
import { NetworkToken } from '../utils/types'
import { useStore } from './useStore'

export const usePricing = (asset?: NetworkToken) => {
  const { currentTokenArrayWithBalance } = useStore()
  const { data } = useQuery(
    ['asset-pricing'],
    () =>
      getSimplePrice(
        [...new Set(Object.values(currentTokenArrayWithBalance).map((a) => a.coingeckoTokenId))].join(',')
      ),
    {
      refetchInterval: 40000,
      staleTime: 30000,
    }
  )

  const getMarketPrice = React.useCallback(
    (assetId: string) => {
      if (data && data[assetId]) {
        const total = Math.round((data[assetId].usd + Number.EPSILON) * 100) / 100
        return utils.commify(total)
      }

      return '0'
    },
    [data]
  )

  const getMarket24HourChange = React.useCallback(
    (assetId: string) => {
      if (data && data[assetId]) {
        return data[assetId].usd_24h_change
      }

      return 0
    },
    [data]
  )

  const getAssetValue = React.useCallback(
    (assetId: string, units: number) => {
      if (data && data[assetId]) {
        const value = Math.round((data[assetId].usd * units + Number.EPSILON) * 100) / 100
        return Math.max(value, 0)
      }

      return 0
    },
    [data]
  )

  const getAssetValueFormatted = React.useCallback(
    (assetId: string, units: number) => {
      return `$${getAssetValue(assetId, units)}`
    },
    [data]
  )
  const getAssetGasValueFormatted = React.useCallback(
    (assetId: string, units: number) => {
      let gasTotal = '0'
      if (data && data[assetId] && asset) {
        gasTotal = (data[assetId].usd * units).toFixed(parseInt(<string>asset?.decimal || '8'))
      }
      return `$${gasTotal}`
    },
    [data]
  )

  return { getMarketPrice, getMarket24HourChange, getAssetValue, getAssetValueFormatted, getAssetGasValueFormatted }
}
