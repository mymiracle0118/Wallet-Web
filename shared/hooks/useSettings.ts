import create from 'zustand'
import { devtools, persist } from 'zustand/middleware'
import produce from 'immer'

import type { Asset } from './useWallet'
import { AccountWallet, useWallet } from './useWallet'
import { SAFETY_MEASURE } from '@portal/portal-extension/src/utils/constants'
import { chromeSyncStorage } from '@portal/shared/utils/chromeSyncStorage'
import { setLockUp } from './useLockUp'
import { EnvironmentType, NetworkToken } from '@portal/shared/utils/types'
import { useStore } from './useStore'

export type AddressBookUser = {
  avatar?: string
  username: string
  address: string
  safety?: SAFETY_MEASURE
  network: string
}
export type AccountTokenType = {
  isFavorite: boolean
}
export type TAccount = {
  assets: Array<Asset>
  address: string
  username: string
  encryptedPrivateKey: string
  encryptedWallet: string
  avatar: string
  balance: number
  isAccountImported: boolean
  networkName: string
  tokens: { [env: string]: { [key: string]: AccountTokenType } }
  customTokens: { [env: string]: { [key: string]: NetworkToken } }
  isPrimary: boolean
}

type TAddressBook = {
  username: string
  address: string
}

export type SettingsState = {
  enablePasswordProtection: boolean
  accounts: { [address: string]: TAccount }
  currentAccount: TAccount | null
  darkMode: boolean
  addressBook: AddressBookUser[]
  saveAccount: (
    username: string,
    address: string,
    networkName: string,
    isPrimary?: boolean,
    isAccountImported?: boolean
  ) => void
  clearAccounts: () => void
  clearAddressbook: () => void
  setDarkMode: (mode: boolean) => void
  setEnablePasswordProtection: (enable: boolean) => void
  hideAccount: (address: string) => void
  addToAddressBook: (user: AddressBookUser) => void
  removeToAddressBook: (address: string) => void
  updateUsername: (username: string, address: string) => void
  networkEnvironment: EnvironmentType
  setNetworkEnvironment: (envType: EnvironmentType) => void
  setCurrentAccount: (account: TAccount) => void
  addToken: (accountKey: string, tokenKey: string) => void
  updateTokens: (accountKey: string, tokenKeys: string[]) => void
  addRemoveFavoriteToken: (accountKey: string, tokenKey: string, isFavoriteValue: boolean) => void
  addCustomToken: (token: NetworkToken) => void
}

export const useSettings = create<SettingsState>()(
  devtools(
    persist(
      (set, get) => ({
        accounts: {},
        currentAccount: null,
        addressBook: [],
        enablePasswordProtection: false,
        darkMode: true,
        networkEnvironment: 'testNet',
        setDarkMode: (mode) => {
          set({ darkMode: mode })
        },

        updateUsername: async (newUsername: string, address: string) => {
          set(
            produce((state: SettingsState) => {
              state.accounts[address].username = newUsername
            })
          )
        },

        addToAddressBook: (user: AddressBookUser) => {
          set(
            produce((state: SettingsState) => ({
              addressBook: [...state.addressBook, user],
            }))
          )
        },

        removeToAddressBook: (address: string) => {
          set(
            produce((state: SettingsState) => {
              state.addressBook = state.addressBook.filter((v) => v.address !== address)
            })
          )
        },

        saveAccount: async (username, address, networkName, isPrimary, isAccountImported) => {
          const { avatar, encryptedWallet } = useWallet.getState()
          const { addTokenToList } = useStore.getState()
          if (username && address) {
            const account = {
              address,
              username,
              encryptedWallet,
              avatar,
              networkName,
              isAccountImported: isAccountImported ? isAccountImported : false,
              tokens: { testNet: {}, mainNet: {} },
              customTokens: { testNet: {}, mainNet: {} },
              isPrimary: isPrimary || false,
            }
            set(
              produce((state: SettingsState) => {
                // @ts-ignore
                state.accounts[address] = account
              })
            )
            const { accounts, currentAccount } = get()
            if (!currentAccount || (currentAccount && !currentAccount.address)) {
              set({ currentAccount: accounts[address] })
            }
            addTokenToList(address, networkName)
          }
        },

        clearAccounts: () => {
          set({ accounts: {}, currentAccount: null })
        },

        clearAddressbook: () => {
          set({ addressBook: [] })
        },

        setEnablePasswordProtection: (enable: boolean) => {
          setLockUp(enable)
          set({
            enablePasswordProtection: enable,
          })
        },

        hideAccount: (addressHide: string) => {
          const accountLists = { ...get().accounts }
          if (accountLists[`${addressHide}`]) {
            delete accountLists[`${addressHide}`]
          }

          set(
            produce((state: SettingsState) => {
              state.accounts = accountLists
            })
          )

          const { walletsList } = useStore.getState()
          if (walletsList[`${addressHide}`]) {
            delete walletsList[`${addressHide}`]
            useStore.getState().removeWalletFromWalletList(walletsList)
          }

          const { address } = useWallet.getState()
          if (address) {
            const account = useSettings.getState().accounts[address]
            useSettings.getState().setCurrentAccount(account)
          }
        },
        setNetworkEnvironment: (envType: EnvironmentType) => {
          set(
            produce((state: SettingsState) => {
              state.networkEnvironment = envType
            })
          )
        },
        setCurrentAccount: (account: TAccount) => {
          const { updateCurrentAccountTokensList } = useStore.getState()
          set(
            produce((state: SettingsState) => {
              state.currentAccount = account
            })
          )
          updateCurrentAccountTokensList(account.address)
        },
        addToken: (accountKey, tokenKey: string) => {
          const { networkEnvironment } = get()
          set(
            produce((state: SettingsState) => {
              const existingToken = { ...state.accounts[accountKey].tokens[networkEnvironment] }
              const newToken: AccountTokenType = { isFavorite: false }
              existingToken[tokenKey] = newToken
              state.accounts[accountKey].tokens[networkEnvironment] = existingToken
            })
          )
        },
        addRemoveFavoriteToken: (accountKey: string, tokenKey: string, isFavoriteValue: boolean) => {
          const { networkEnvironment } = get()
          set(
            produce((state: SettingsState) => {
              const existingToken = { ...state.accounts[accountKey].tokens[networkEnvironment] }
              const newToken: AccountTokenType = { isFavorite: false }
              existingToken[tokenKey] = { ...existingToken[tokenKey], isFavorite: isFavoriteValue }
              state.accounts[accountKey].tokens[networkEnvironment] = existingToken
            })
          )
        },
        updateTokens: (accountKey: string, tokenKeys: string[]) => {
          const { networkEnvironment } = get()
          set(
            produce((state: SettingsState) => {
              const existingTokens = { ...state.accounts[accountKey].tokens[networkEnvironment] }
              for (const key in existingTokens) {
                if (!tokenKeys.includes(key)) {
                  delete existingTokens[key]
                }
              }
              state.accounts[accountKey].tokens[networkEnvironment] = existingTokens
            })
          )
        },
        addCustomToken: async (token: NetworkToken) => {
          const { networkEnvironment, currentAccount } = get()
          if (currentAccount) {
            const accountKey = currentAccount.address
            set(
              produce((state: SettingsState) => {
                if (!state.accounts[accountKey].customTokens[networkEnvironment]) {
                  state.accounts[accountKey].customTokens[networkEnvironment] = {}
                }
                const existingToken = { ...state.accounts[accountKey].customTokens[networkEnvironment] }
                existingToken[token.shortName] = token
                state.accounts[accountKey].customTokens[networkEnvironment] = existingToken
              })
            )

            const { addTokenToList, addWalletToList, walletsList } = useStore.getState()

            const newWalletObj: AccountWallet = { ...walletsList[accountKey]['ETH'] }
            await addWalletToList(accountKey, token.networkName, newWalletObj)
            addTokenToList(accountKey, token.shortName)
          }
        },
      }),
      {
        name: 'account-storage',
        getStorage: () => chromeSyncStorage,
      }
    ),
    { name: 'Shuttle - Accounts' }
  )
)
