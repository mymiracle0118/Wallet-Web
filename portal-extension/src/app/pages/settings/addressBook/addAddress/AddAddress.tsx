import { Menu } from '@headlessui/react'
import { yupResolver } from '@hookform/resolvers/yup'
import { default as networkAssets } from '@portal/shared/data/networks.json'
import { useSettings } from '@portal/shared/hooks/useSettings'
import { listNetwork } from '@portal/shared/hooks/useWallet'
import { CloseRoundedIcon } from '@src/app/components/Icons'
import { Button, CustomTypography, Dropdown, DropdownItem, Form, Icon, Input } from 'components'
import SinglePageTitleLayout from 'layouts/single-page-layout/SinglePageLayout'
import { goBack } from 'lib/woozie'
import { useCallback, useState } from 'react'
import { useForm } from 'react-hook-form'
import { useTranslation } from 'react-i18next'
import * as yup from 'yup'
import DropDownIcon from '../../../../../../public/images/backgrounds/dropdown_icon.svg'

const schema = yup.object().shape({
  address: yup.string().required(),
  label: yup.string().required(),
})

const AddAddress = () => {
  const { t } = useTranslation()
  const { addToAddressBook } = useSettings()
  const [addressValue, setAddressValue] = useState<string>('')
  const [activeNetwork, setActiveNetwork] = useState<string>(networkAssets[0].networkId)

  const methods = useForm({
    resolver: yupResolver(schema),
    mode: 'onChange',
  })
  const {
    handleSubmit,
    formState: { errors, isValid, isDirty },
    register,
  } = methods

  const addAddress = useCallback(
    (data: { label: string; address: string }) => {
      addToAddressBook({
        address: data.address,
        username: data.label,
        network: activeNetwork,
      })
      goBack()
      setAddressValue('')
    },
    [addToAddressBook, activeNetwork]
  )

  const handleChangeNetwork = (val: string) => {
    const selected = networkAssets.find((v) => v.name === val) as listNetwork
    setActiveNetwork(selected.networkId)
  }

  return (
    <SinglePageTitleLayout title={t('Settings.addAddressTitle')}>
      <Form methods={methods} onSubmit={handleSubmit(addAddress)}>
        <div className="space-y-4">
          <CustomTypography type="secondary">{t('Settings.addAddressSubTitle')}</CustomTypography>

          <Dropdown
            classDynamicChild="h-[16.25rem] overflow-x-hidden overflow-y-scroll w-full border border-solid border-[#424250]"
            classDynamicMenu="bg-custom-white10 mb-4 mt-6 rounded-md p-1 !table"
            anchor={
              <Menu.Button data-aid="currencyDropdown" className="p-2 rounded-xl flex items-center gap-2">
                <img alt="icon" src={activeNetwork?.image} className="h-6 rounded-full" />

                <CustomTypography variant="subtitle" className="w-60 text-left mr-4">
                  {activeNetwork?.name}
                </CustomTypography>
                <Icon size="small" icon={<DropDownIcon />} />
              </Menu.Button>
            }
          >
            {networkAssets.map((network) => (
              <DropdownItem
                key={network.name}
                active={network.name === activeNetwork?.name}
                text={network.name}
                icon={network.image}
                isImg={true}
                onSelect={handleChangeNetwork}
              />
            ))}
          </Dropdown>

          <Input
            placeholder="Address"
            mainColor
            fullWidth
            dataTestId="addAddress-address"
            dataAid="address"
            id="address"
            {...register('address')}
            value={addressValue}
            onChange={(e) => setAddressValue(e.target.value)}
            error={errors.address?.message}
            endAdornment={
              <Button
                variant="flat"
                size="sm"
                color="transparent"
                aria-label="toggle password visibility"
                className="bg-transparent"
                isIconOnly
                onClick={() => setAddressValue('')}
              >
                <CloseRoundedIcon />
              </Button>
            }
          />
          <Input
            placeholder="Label"
            mainColor
            fullWidth
            className="mt-2"
            dataTestId="addAddress-label"
            dataAid="label"
            id="label"
            {...register('label')}
            error={errors.label?.message}
          />
          <div className="flex gap-2 pt-4">
            <Button variant="bordered" color="outlined" onClick={goBack}>
              {t('Actions.cancel')}
            </Button>
            <Button
              type="submit"
              color={`${!isValid || !isDirty ? 'disabled' : 'primary'}`}
              isDisabled={!isValid || !isDirty}
              data-test-id="add-address-btn"
            >
              {t('Actions.add')}
            </Button>
          </div>
        </div>
      </Form>
    </SinglePageTitleLayout>
  )
}

export default AddAddress
