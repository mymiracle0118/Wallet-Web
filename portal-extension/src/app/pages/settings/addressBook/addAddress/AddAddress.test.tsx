import React from 'react'
import { FormProvider, useForm } from 'react-hook-form'
import { render, renderHook, screen } from '@testing-library/react'
import { useSettings } from '@portal/shared/hooks/useSettings'
import userEvent from '@testing-library/user-event'

import AddAddress from './AddAddress'

const WrapperForm = ({ children }) => {
  const methods = useForm()
  return <FormProvider {...methods}>{children}</FormProvider>
}
describe('<AddAddress />', () => {
  test('renders correctly', () => {
    const tree = render(
      <WrapperForm>
        <AddAddress />
      </WrapperForm>
    )

    expect(tree).toMatchSnapshot()
  })

  test('Add item to Address Book', async () => {
    const user = userEvent.setup()
    const { result } = renderHook(() => useSettings())
    const before = result.current.addressBook.length
    render(
      <WrapperForm>
        <AddAddress />
      </WrapperForm>
    )

    await user.type(screen.getByTestId('addAddress-address'), '0xb81B777a6f7C8d79400A5172c5316F2233dd76C4')
    await user.type(screen.getByTestId('addAddress-label'), 'Test Address 1')
    await user.click(screen.getByTestId('add-address-btn'))

    expect(result.current.addressBook).toHaveLength(before + 1)
  })
})
