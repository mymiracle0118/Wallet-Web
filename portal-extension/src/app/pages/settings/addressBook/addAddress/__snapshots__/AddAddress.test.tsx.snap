// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddAddress /> renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="dark:bg-surface-dark bg-custom-white flex flex-col h-[600px] dark:text-custom-white text-custom-black"
      >
        <div
          class="flex flex-row relative justify-between items-center h-16"
        >
          <button
            class="h-[24px] w-[24px] p-1 ml-1 hover:bg-custom-grey10 dark:hover:bg-custom-white10 rounded-full flex items-center justify-center absolute"
            type="button"
          >
            <arrow-left.svg
              classname="text-xl stroke-custom-black dark:stroke-custom-white"
            />
          </button>
          <div
            class="flex-1 text-center"
          >
            <div
              class="text-custom-black dark:text-custom-white text-lg font-extrabold"
            >
              Settings.addAddressTitle
            </div>
          </div>
        </div>
        <div
          class="w-full h-full overflow-auto"
        >
          <form>
             
            <div
              class="px-4"
            >
              <div
                class="text-custom-grey100 dark:text-custom-white40"
                style="text-align: left; margin-bottom: 16px;"
              >
                Settings.addAddressSubTitle
              </div>
              <div
                class="flex relative rounded-md items-center hover:shadow-md w-full bg-input-light-bg dark:bg-custom-white10"
              >
                <input
                  autocomplete="off"
                  class="outline-none	py-3.5 px-3.5 w-full text-sm text-custom-black dark:text-custom-white bg-transparent bg-transparent text-h4 "
                  data-aid="address"
                  data-testid="addAddress-address"
                  id="address"
                  name="address"
                  placeholder="Address"
                  value=""
                />
                <div
                  class="pr-3.5"
                >
                  <div
                    class="MuiInputAdornment-root MuiInputAdornment-positionEnd css-1laqsz7-MuiInputAdornment-root"
                  >
                    <button
                      aria-label="toggle password visibility"
                      class="rounded-full p-2 dark:hover:bg-custom-white40 hover:bg-custom-grey40"
                      type="button"
                    >
                      <x-circle.svg
                        classname="stroke-custom-black dark:stroke-custom-white"
                      />
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="flex relative rounded-md items-center hover:shadow-md w-full bg-input-light-bg dark:bg-custom-white10 mt-2"
              >
                <input
                  autocomplete="off"
                  class="outline-none	py-3.5 px-3.5 w-full text-sm text-custom-black dark:text-custom-white bg-transparent bg-transparent text-h4 "
                  data-aid="label"
                  data-testid="addAddress-label"
                  id="label"
                  name="label"
                  placeholder="Label"
                  value=""
                />
              </div>
              <div
                class="flex gap-2 mt-3"
              >
                <button
                  class="flex items-center justify-center rounded-md hover:cursor-pointer font-extrabold text-sm w-full h-[44px] border-[1px] border-solid text-primary dark:text-custom-white border-primary dark:border-custom-white40 dark:hover:border-custom-white hover:border-primary-alt hover:text-primary-alt"
                  type="button"
                >
                  Actions.cancel
                </button>
                <button
                  class="flex items-center justify-center rounded-md hover:cursor-pointer font-extrabold text-sm w-full disabled:text-custom-grey40 dark:disabled:text-custom-grey40 h-[44px] text-custom-white dark:text-custom-white bg-custom-grey10"
                  data-testid="add-address-btn"
                  disabled=""
                  type="submit"
                >
                  Actions.add
                </button>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="dark:bg-surface-dark bg-custom-white flex flex-col h-[600px] dark:text-custom-white text-custom-black"
    >
      <div
        class="flex flex-row relative justify-between items-center h-16"
      >
        <button
          class="h-[24px] w-[24px] p-1 ml-1 hover:bg-custom-grey10 dark:hover:bg-custom-white10 rounded-full flex items-center justify-center absolute"
          type="button"
        >
          <arrow-left.svg
            classname="text-xl stroke-custom-black dark:stroke-custom-white"
          />
        </button>
        <div
          class="flex-1 text-center"
        >
          <div
            class="text-custom-black dark:text-custom-white text-lg font-extrabold"
          >
            Settings.addAddressTitle
          </div>
        </div>
      </div>
      <div
        class="w-full h-full overflow-auto"
      >
        <form>
           
          <div
            class="px-4"
          >
            <div
              class="text-custom-grey100 dark:text-custom-white40"
              style="text-align: left; margin-bottom: 16px;"
            >
              Settings.addAddressSubTitle
            </div>
            <div
              class="flex relative rounded-md items-center hover:shadow-md w-full bg-input-light-bg dark:bg-custom-white10"
            >
              <input
                autocomplete="off"
                class="outline-none	py-3.5 px-3.5 w-full text-sm text-custom-black dark:text-custom-white bg-transparent bg-transparent text-h4 "
                data-aid="address"
                data-testid="addAddress-address"
                id="address"
                name="address"
                placeholder="Address"
                value=""
              />
              <div
                class="pr-3.5"
              >
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionEnd css-1laqsz7-MuiInputAdornment-root"
                >
                  <button
                    aria-label="toggle password visibility"
                    class="rounded-full p-2 dark:hover:bg-custom-white40 hover:bg-custom-grey40"
                    type="button"
                  >
                    <x-circle.svg
                      classname="stroke-custom-black dark:stroke-custom-white"
                    />
                  </button>
                </div>
              </div>
            </div>
            <div
              class="flex relative rounded-md items-center hover:shadow-md w-full bg-input-light-bg dark:bg-custom-white10 mt-2"
            >
              <input
                autocomplete="off"
                class="outline-none	py-3.5 px-3.5 w-full text-sm text-custom-black dark:text-custom-white bg-transparent bg-transparent text-h4 "
                data-aid="label"
                data-testid="addAddress-label"
                id="label"
                name="label"
                placeholder="Label"
                value=""
              />
            </div>
            <div
              class="flex gap-2 mt-3"
            >
              <button
                class="flex items-center justify-center rounded-md hover:cursor-pointer font-extrabold text-sm w-full h-[44px] border-[1px] border-solid text-primary dark:text-custom-white border-primary dark:border-custom-white40 dark:hover:border-custom-white hover:border-primary-alt hover:text-primary-alt"
                type="button"
              >
                Actions.cancel
              </button>
              <button
                class="flex items-center justify-center rounded-md hover:cursor-pointer font-extrabold text-sm w-full disabled:text-custom-grey40 dark:disabled:text-custom-grey40 h-[44px] text-custom-white dark:text-custom-white bg-custom-grey10"
                data-testid="add-address-btn"
                disabled=""
                type="submit"
              >
                Actions.add
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
